Introduce CleaningEvent to the codebase.

6ar8nas opened the pull request:
## What was done

- Model `CleaningEvent` and some of its public models created, created a mapping profile for cleaning events.
- Added `CleaningEvent` to the `PollutedLocationContext` (to do this as advised [in most cases](https://learn.microsoft.com/en-us/ef/core/modeling/relationships) I added `List<CleaningEvent>` to the `PollutedLocation` class, which kind of also seems to make sense. All of related public models, mappings and validators adjusted too.
- Created `CleaningEventDatabaseRepository` and implemented most of its actions (update missing, as I did it based on the current repository of `PollutedLocation`s).
- Added some events to the DbInitializer. 
- Created a `CleaningEventController`, with GetById, GetAll
- Tests written
- Covered some previous technical debt - wrote more integration and unit tests, adjusted try/catch statements to only be around where we call db methods and etc.

<!-- If needed, describe your changes in detail here -->

This PR will allow more than one person to work in parallel, as the base for this model will already be up and running.

Closes #113 

Sorry for the big PR. Didn't expect to do it all in one, but one thing lead to another and this happened :D

Related [PR#42](https://github.com/vu-vibedosa/apsitvarkom-fe/pull/42) in front-end.

6ar8nas committed:
Apsitvarkom.Models project part of the implementation covered

6ar8nas committed:
introduced TidyingEvents to DataAccess project.

6ar8nas committed:
introduced a controller

6ar8nas committed:
fix failing due to property mapping not being determined

6ar8nas committed:
minor implementation changes

6ar8nas committed:
unit and integration testing additions

6ar8nas commented:
Not everything is correct with DbInitializer and PollutedLocationContext (probably OnModelCreating part [?]).
Will delve into it more later.

6ar8nas committed:
minor change from new Guid to Guid.Parse

6ar8nas committed:
readjusted dbInitializer to now return the same instance as added to db to TidyingEvents table.

6ar8nas commented:
I'm missing something regarding Events being returned from database. They seem to exist in the db, but they're not returned from it. Will look into it tomorrow.

6ar8nas committed:
resolved failing tests issue

6ar8nas commented on Apsitvarkom.Models/PollutedLocation.cs:
```diff
@@ -32,4 +32,7 @@ public enum SeverityLevel
     public Location Location { get; set; } = new();
+
+    /// <summary>Collection of events linked to the location.</summary>
+    public List<TidyingEvent> Events { get; set; } = new();
```
Apparently, DB doesn't return any data for this `Events` property by default, as it is (logically, due to unwanted data repetitions) not stored in the `PollutedLocation` table. This is why I didn't include it in the `PollutedLocationResponse`. Will play around and see if we could possibly return it anyways, as I believe it would be beneficial when listing all the events in the sidebar as well.

6ar8nas replied:
resolved in `PollutedLocationDatabaseRepository`, where I had to specify that I'd like to include `Events` into the returned `PollutedLocation` instance(s).

BenasB replied:
Yes, you have to tell EF core explicitly to include nested complex types

6ar8nas committed:
some cleanup done

6ar8nas committed:
changed the implementation to now include Events in the PollutedLocationResponse
Created TidyingEventResponse

6ar8nas committed:
TidyingEvent get actions implemented and covered with tests

6ar8nas committed:
added one integration test, async/await for in-memory db calls.

6ar8nas commented on Apsitvarkom.DataAccess/PollutedLocationContext.cs:
```diff
@@ -6,16 +6,18 @@ namespace Apsitvarkom.DataAccess;
     public PollutedLocationContext(DbContextOptions<PollutedLocationContext> options) : base(options) { }
-
+    
     protected override void OnModelCreating(ModelBuilder modelBuilder)
```
Didn't figure out a way to not allow creation of `TidyingEvent` if there is no such `PollutedLocation` yet.
Probably one way to achieve it could be introducing a `PollutedLocation` property inside of the `TidyingEvent` model and then specifying `modelBuilder.Entity<TidyingEvent>.Property(x => x.PollutedLocation).IsRequired();`

6ar8nas replied:
I've already tried 
`modelBuilder.Entity<TidyingEvent>().HasOne<PollutedLocation>().WithMany(l => l.Events).HasForeignKey(e => e.PollutedLocationId).IsRequired();` - doesn't seem to work

6ar8nas replied:
another workaround could be just checking if a PollutedLocation by the given id exists in controller, I suppose

BenasB replied:
> Probably one way to achieve it could be introducing a PollutedLocation property inside of the TidyingEvent model

And that's exactly what you should do IMO, as per [Fully defined relationships](https://learn.microsoft.com/en-us/ef/core/modeling/relationships?tabs=fluent-api%2Cfluent-api-simple-key%2Csimple-key#fully-defined-relationships) docs, they have the same exact setup: "Blog" is our "PollutedLocation" and "Post" is our "TidyingEvent". We have a 1 (polluted location) to N (many TidyingEvents) relation

6ar8nas committed:
validation tests extended, slighty modified

6ar8nas committed:
integration tests for IPollutedLocationContext extended

6ar8nas committed:
integration tests slightly readjusted

6ar8nas committed:
unit tests covering try/catch statements for controllers added

6ar8nas commented on Apsitvarkom.Models/Public/TidyingEventResponse.cs:
```diff
@@ -0,0 +1,12 @@
+    public DateTime StartTime { get; set; }
+
+    public string? Notes { get; set; }
+}
```
Didn't think there was any need for a validator here, hence there is none, which is different to what we have in most of the other public models.

BenasB replied:
Response validators are used to document swagger (show schema), for example that the `Id` is not nullable

BenasB replied:
Nevermind, nullable is determined automatically, these validators to document swagger with **business rules**, but this model doesn't really have any right now

BenasB commented on Apsitvarkom.Api/Program.cs:
```diff
@@ -33,6 +34,7 @@
 
 builder.Services
     .AddAutoMapper(typeof(PollutedLocationProfile))
+    .AddAutoMapper(typeof(TidyingEventProfile))
```
Do you need this additional `AddAutoMapper`? I think that since both of our Profiles are in the same [assembly](https://learn.microsoft.com/en-us/dotnet/standard/assembly/), we only need to specify one.

> And then you let AutoMapper know in what **assemblies** are those profiles...

Ref: https://docs.automapper.org/en/stable/Dependency-injection.html

6ar8nas committed:
Merge branch 'staging' into TidyingEventBaseImplementation

BenasB commented on Apsitvarkom.DataAccess/PollutedLocationDatabaseRepository.cs:
```diff
@@ -21,21 +21,25 @@ public PollutedLocationDatabaseRepository(IPollutedLocationContext context)
     /// <inheritdoc />
-    public Task<PollutedLocation?> GetByPropertyAsync(Expression<Func<PollutedLocation, bool>> propertyCondition)
+    public async Task<PollutedLocation?> GetByPropertyAsync(Expression<Func<PollutedLocation, bool>> propertyCondition)
     {
-        return _context.PollutedLocations.FirstOrDefaultAsync(propertyCondition);
+        return await _context.PollutedLocations
+            .Include(l => l.Events)
+            .FirstOrDefaultAsync(propertyCondition);
```
`async` `await` is not needed here, it just adds overhead, see https://stackoverflow.com/a/38017167/13135665

As a rule of thumb, if you can directly return a `Task`, you should. Might be worth checking other methods also

BenasB commented on Apsitvarkom.DataAccess/TidyingEventDatabaseRepository.cs:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    /// <inheritdoc />
+    public async Task<TidyingEvent?> GetByPropertyAsync(Expression<Func<TidyingEvent, bool>> propertyCondition)
+    {
+        return await _context.TidyingEvents.FirstOrDefaultAsync(propertyCondition);
+    }
```
Here also for example (referring to comment about `async` `await`)

6ar8nas committed:
PollutedLocation Delete action adjustments - StatusCode204 instead of 200 (#115)

Implemented Repository method ExistsByPropertyAsync
Changed Delete action in PollutedLocationController
Wrote tests to cover the implementation
Renamed TidyingEvent to CleaningEvent

6ar8nas committed:
removed mapper
removed overhead from excessive async await instead of simply returning Task

6ar8nas committed:
added PollutedLocation to CleaningEvent to define the 1-M relationship fully.

BenasB committed:
Fix integration tests

BenasB committed:
Fix tests .Value
